### PR TITLE
Fix 5716: Enabling Solana

### LINKED ISSUE #5716: Wallet: Enable Solana Support for iOS
### Description: 
Once Solana Support is enabled, users can create or import solana account, switch to solana networks, fetch solana account balance and make send transactions with a solana account

### PR DESCRIPTION
**Pull request by nuo-xu**
## Summary of Changes

This pull request fixes #5716 

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [ ] User-facing strings use `NSLocalizableString()`
- [ ] New or updated UI has been tested across:
  - [ ] Light & dark mode
  - [ ] Different size classes (iPhone, landscape, iPad)
  - [ ] Different dynamic type sizes

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

**Review by iccub — APPROVED** [review]

File: BraveWallet/WalletConstants.swift [hunk 1]
```diff
@@ -47,5 +47,5 @@ struct WalletConstants {
 }
 
 struct WalletDebugFlags {
-  static var isSolanaEnabled: Bool = false
+  static var isSolanaEnabled: Bool = true
```
Comment (iccub) [inline BraveWallet/WalletConstants.swift]: We could remove this flag altogether maybe, sometime in the future

Comment (StephenHeaps) [reply BraveWallet/WalletConstants.swift]: I agree, I would suggest dropping this flag entirely (and just include `.sol` in `supportedCoinTypes` array) would be good in this PR.

Comment (nuo-xu) [reply BraveWallet/WalletConstants.swift]: removed this flag in [5a674dd](https://github.com/brave/brave-ios/pull/5719/commits/5a674ddab0e3e5b440cc259786a44e874b8fea3a)
re-request the team to review.

**Review by iccub — APPROVED** [review]

**Comment by kdenhartog** [comment]
Spotted a minor UI bug
receiving address error bug

**Review by StephenHeaps — APPROVED** [review]

**Comment by nuo-xu** [comment]
hey @kdenhartog, we created issue [5773](https://github.com/brave/brave-ios/issues/5773) and it has been fixed in `development`. It fixes the `same address` error message didn't update after selected account changed. We couldn't reproduce the `invalid SOL address` error with the account `AX1fumgqvHhXe2gRAJiQK2d759D4LPXaWjEVDRHeoG6B`. If you can reproduce, please provide us more information. thanks!

**Comment by kdenhartog** [comment]
Ahh sorry @nuo-xu, yeah that was the only bug I was trying to report was the `same address` error. The `invalid SOL address` error was from me double pasting the address in. Looks like you got it all fixed

**Comment by kdenhartog** [comment]
**Summary: I think if we can get these two issues fixed this would be good from my perspective. Will let @stoletheminerals comment on their findings as well**

I've tried the following:

- Rejecting transactions (works as expected)
- Submitting normal transaction (works as expected)
- Submitting to program address (submits to wallet, but chain errors and no funds lost)
- playing around with bad inputs for account porting

---

Issues found:
1. Doesn't appear to be properly detecting all assets in the wallet. When I manually added the asset as viewable then it shows up, but it doesn't show the value properly either.
Screen Shot 2022-08-04 at 6 30 38 PM

---

2. Can add the same account twice. Once as a primary account and once as a secondary account. Steps to repro:
  - copy a private key in a different wallet on your HD path (e.g. account 2) (I used slope finance to export the key)
  - import the key into the wallet by pasting the private key in
  - add an account again without pasting anything in

Screen Shot 2022-08-04 at 6 45 46 PM

**Comment by stoletheminerals** [comment]
Should we put the SOL icon here instead of a basic grey circle? It would be more apparent to the user which asset is selected at the moment. Ideally a colour icon for the mainet and greyscale for test/dev.
image
image

**Comment by nuo-xu** [comment]
@stoletheminerals the icon issue has been fixed in `development` ref: https://github.com/brave/brave-ios/pull/5772

**Comment by darkdh** [comment]
>Can add the same account twice. Once as a primary account and once as a secondary account. Steps to repro:

@kdenhartog this is not iOS specific issue. The thing you are trying to do is import an account which hasn't been derived and we can't derived all possible accounts just to check if the key you are trying to import belongs to derived accounts. We only check what users have derived. (https://github.com/brave/brave-core/blob/2699c59acec157de3fd929b7d922b0c2d2e6beca/components/brave_wallet/browser/hd_keyring.cc#L63-L70)

The possible solution would be, when deriving account we check if the account being derived is in import accounts, if so delete the import account. But it will slow down derivation process when import accounts grew.

**Comment by stoletheminerals** [comment]
Overall looks good to me!
Just a note: Since we don't support ENS / SOL domains on iOS I think we should remove the "URL" part from the "To" field on both Solana and Ethereum windows for now.

**Comment by kdenhartog** [comment]
> @kdenhartog this is not iOS specific issue. The thing you are trying to do is import an account which hasn't been derived and we can't derived all possible accounts just to check if the key you are trying to import belongs to derived accounts. We only check what users have derived. ([brave/brave-core@2699c59/components/brave_wallet/browser/hd_keyring.cc#L63-L70](https://github.com/brave/brave-core/blob/2699c59acec157de3fd929b7d922b0c2d2e6beca/components/brave_wallet/browser/hd_keyring.cc#L63-L70))

Ok thanks for clarifying @darkdh that makes sense to me why this is possible then. I can see how that lookup would have tradeoffs here as the derivation gets larger so at this point, given I couldn't produce any unintended side effects through this, I'm good to leave this as is.

**Review by kdenhartog — APPROVED** [review]
Issue 1 from https://github.com/brave/brave-ios/pull/5719#issuecomment-1204830858 isn't a security vulnerability so not going to block the review on this. Approving I think this one is good to go from a sec perspective

**Comment by stoletheminerals** [comment]
Just noticed: I can't choose my second Solana account on Send window:
image
Clicking the pointing down arrow should let me choose my second Solana account, but instead, I can only copy the address of my first account.
UPD: I can't switch Ethereum accounts as well

**Comment by StephenHeaps** [comment]
> Just noticed: I can't choose my second Solana account on Send window:
> image
> Clicking the pointing down arrow should let me choose my second Solana account, but instead, I can only copy the address of my first account.
> UPD: I can't switch Ethereum accounts as well

@stoletheminerals are you on iOS 16 beta? If so it’s a known issue:
https://github.com/brave/brave-ios/issues/5618

Hopefully can have this fixed soon.

**Comment by stoletheminerals** [comment]
@StephenHeaps yep, on 16 beta. Didn't see that issue, my bad!

**Comment by nuo-xu** [comment]
fyi @kdenhartog issue 1* https://github.com/brave/brave-ios/issues/5779 is also resolved.

**Pull request merged**